### PR TITLE
ix: resholve so we do not have to wrap

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Description of changes

resholve is *neat*!

Cc: @asymmetric @artturin

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/ix/default.nix [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
-  postFixup = ''
-    wrapProgram $out/bin/ix --prefix PATH : "${lib.makeBinPath [ curl ]}"
+    runHook postInstall
   '';
 
+  solutions.default = {
+    scripts = [ "bin/ix" ];
+    interpreter = runtimeShell;
+    inputs = [ curl ];
+    keep."$echo" = true;
```
Comment (abathur) [inline pkgs/tools/misc/ix/default.nix:26]: It looks like this is solving for this line in the target script:

```bash
            d) $echo curl $opts -X DELETE ix.io/$OPTARG; return;;
```

I don't see an `echo` variable defined anywhere in the script, so I wonder if this is a bug/typo in the source? @peterhoeg @asymmetric have either of you used this `-i` flag?

**Comment by asymmetric** [comment]
@peterhoeg what's the benefit of using `resholve` instead of wrapping?

Also more generally, if we do it for `ix`, shouldn't we do it for lots of other packages? Then the issue becomes, why not go through an RFC rather than fragmenting nixpkgs further, with no explanation of why each approach is chosen?

**Comment by abathur** [comment]
> @peterhoeg what's the benefit of using `resholve` instead of wrapping?

Peter may have distinct reasons, but in a narrow case like this I'd cite two:

1. A wrapped script has per-execution overhead. I'm not familiar with `ix`, so I don't know if it's large enough to be meaningful or not. For example:

     ```console
    $ ./ix -d 44VS
    Wed Jul 20 2022 10:03:31 --> 
    user required
    Wed Jul 20 2022 10:03:31 (195ms)

    $ ./ix -d 44VS
    Wed Jul 20 2022 10:03:34 --> 
    user required
    Wed Jul 20 2022 10:03:34 (189ms)

    $ ./ix -d 44VS
    Wed Jul 20 2022 10:03:36 --> 
    user required
    Wed Jul 20 2022 10:03:37 (186ms)

    $ ix -d 44VS
    Wed Jul 20 2022 10:03:39 --> 
    user required
    Wed Jul 20 2022 10:03:39 (261ms)

    $ ix -d 44VS
    Wed Jul 20 2022 10:03:41 --> 
    user required
    Wed Jul 20 2022 10:03:41 (265ms)

    $ ix -d 44VS
    Wed Jul 20 2022 10:03:43 --> 
    user required
    Wed Jul 20 2022 10:03:43 (268ms)
    ```

2. resholve blocks the build when it finds commands it doesn't have an executable for, so it's helping ensure everything is present and remains present even if the source gets updated to add an extra dependency someday. (I don't really expect that would happen with `ix` given its changelog, but it's still providing the assurance and backstop.)

> Also more generally, if we do it for `ix`, shouldn't we do it for lots of other packages? Then the issue becomes, why not go through an RFC rather than fragmenting nixpkgs further, with no explanation of why each approach is chosen?

I'm a little unsure from the questions, so it's probably good to make sure we're on the same page. `resholve` is providing an API for packaging some Bash/Shell projects, and isn't a general replacement for makeWrapper. It is one of many ways to deal with shell script dependencies alongside options like inlining the script with interpolation, using sed or one of the substitute functions, etc.

**Comment by peterhoeg** [comment]
ix might be the most simple script we are in fact packaging in nixpkgs, so the overhead issue doesn’t really apply here. I must admit I have become a bit of a resholve fanboy because it’s elegant and provides strong guarantees that we have in fact declared all our inputs correctly.

Should we start doing this across the board? Well, an argument could definitely be made for that, but I think it’s important to get a few of the scripts converted over to actually prove that it works and *then* we can have the conversation about what happens next.

File: pkgs/tools/misc/ix/default.nix (hunk 1 above)
Comment (peterhoeg) [reply pkgs/tools/misc/ix/default.nix]: I've patched it out.

File: pkgs/tools/misc/ix/default.nix [hunk 2]
```diff
@@ -15,15 +15,16 @@ resholve.mkDerivation {
...
+    substituteInPlace $out/bin/ix \
+      --replace '$echo ' ""
 
     runHook postInstall
   '';
 
   solutions.default = {
     scripts = [ "bin/ix" ];
-    interpreter = runtimeShell;
+    interpreter = "${lib.getBin bash}/bin/bash";
```
Comment (Artturin) [inline pkgs/tools/misc/ix/default.nix:26]:
```suggestion
    interpreter = "${lib.getExe bash}";
```

Comment (peterhoeg) [reply pkgs/tools/misc/ix/default.nix]: I ididn't know getExe, thanks!

**Comment by peterhoeg** [comment]
I think all technical issues have been addressed, but we obviously shouldn't merge this without @asymmetric's approval as maintainer. Any thoughts, concerns, reservations?

**Comment by asymmetric** [comment]
@peterhoeg I'm OK with this as long as you take over maintaining the package, i.e. remove me and add yourself :)

**Comment by asymmetric** [comment]
My reservations are that it's yet another system to learn, for no obvious benefit in this case. I don't have the time to look into how resholve (and its dependencies) work.

**Comment by peterhoeg** [comment]
Will merge this after the checks pass.

> I don't have the time to look into how resholve (and its dependencies) work.

Valid point. I can however not recommend it strongly enough (and no, @abathur isn't paying me a commission). I have slowly been moving my own stuff over when I needed to touch the scripts anyway and this very weekend, it caught a missing dependency for me at time of deployment rather than runtime.

**Comment by Artturin** [comment]
afaik the common consensus is that wrappers are a hack

**Pull request merged**